librz/socket/socket.c

```diff
@@ -898,4 +912,160 @@ RZ_API ut8 *rz_socket_slurp(RzSocket *s, int *len) {
 	return buf;
 }
 
+struct rz_stop_pipe_t {
+#if __WINDOWS__
+	WSAEVENT event;
+#else
+	int fds[2];
+#endif
+}; // RzStopPipe
+
+RZ_API RzStopPipe *rz_stop_pipe_new(void) {
+	RzStopPipe *stop_pipe = RZ_NEW0(RzStopPipe);
+	if (!stop_pipe) {
+		return NULL;
+	}
+#if __WINDOWS__
+	stop_pipe->event = WSACreateEvent();
+	if (stop_pipe->event == WSA_INVALID_EVENT) {
+		free(stop_pipe);
+		return NULL;
+	}
+#else
+	int r = pipe(stop_pipe->fds);
+	if (r < 0) {
+		free(stop_pipe);
+		return NULL;
+	}
+	r = fcntl(stop_pipe->fds[0], F_SETFL, O_NONBLOCK);
+	if (r == -1) {
+		close(stop_pipe->fds[0]);
+		close(stop_pipe->fds[1]);
+		free(stop_pipe);
+		return NULL;
+	}
+#endif
+	return stop_pipe;
+}
+
+RZ_API void rz_stop_pipe_free(RzStopPipe *stop_pipe) {
+	if (!stop_pipe) {
+		return;
+	}
+#if __WINDOWS__
+	WSACloseEvent(stop_pipe->event);
+#else
+	close(stop_pipe->fds[0]);
+	close(stop_pipe->fds[1]);
+#endif
+	free(stop_pipe);
+}
+
+/**
+ * Trigger a stop on the stop pipe, causing rz_stop_pipe_select_single() on
+ * it to return with RZ_STOP_PIPE_STOPPED.
+ * This function is async-signal-safe.
+ */
+RZ_API void rz_stop_pipe_stop(RzStopPipe *stop_pipe) {
+	// No rz_return_if_fail because that would not be async-signal-safe!
+#if __WINDOWS__
+	WSASetEvent(stop_pipe->event);
+#else
+	(void)!write(stop_pipe->fds[1], "\x00", 1); // no print on err for async-signal-safety
+#endif
+}
+
+/**
+ * Block until one of the following events has happened and return:
+ * * \p stop_pipe has been stopped with `rz_stop_pipe_stop()` => `RZ_STOP_PIPE_STOPPED`
+ * * \p sock is ready for read or write (depending on \p sock_write) => `RZ_STOP_PIPE_SOCKET_READY`
+ * * \p timeout_ms milliseconds have passed => `RZ_STOP_PIPE_TIMEOUT`
+ *
+ * This is only supported for non-ssl sockets.
```

Comment:
why?

Reply:
rz_socket just supports TLS why not just add a eval for this?

Reply:
The select operates on the raw `sock->fd`. When ssl is used in the `RzSocket`, then it does not have that and instead uses OpenSSL functions on an `s->sfd`.